util/files/tar.go

```diff
@@ -0,0 +1,169 @@
+package files
+
+import (
+	"archive/tar"
+	"bufio"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type tgz struct {
+	srcPath    string
+	exclusions []string
+	tarWriter  *tar.Writer
+}
+
+// Tgz will iterate over all files found in srcPath compressing them with gzip
+// and archiving with Tar. Will invoke every given writer while generating the tgz.
+// This is useful to generate checksums. Will exclude files matching the exclusions
+// list blob.
+func Tgz(srcPath string, exclusions []string, writers ...io.Writer) error {
+	if _, err := os.Stat(srcPath); err != nil {
+		return fmt.Errorf("error inspecting srcPath %q: %w", srcPath, err)
+	}
+
+	mw := io.MultiWriter(writers...)
+
+	gzw := gzip.NewWriter(mw)
+	defer gzw.Close()
+
+	tw := tar.NewWriter(gzw)
+	defer tw.Close()
+
+	t := &tgz{
+		srcPath:    srcPath,
+		exclusions: exclusions,
+		tarWriter:  tw,
+	}
+	return filepath.Walk(srcPath, t.tgzFile)
+}
+
+// Untgz will loop over the tar reader creating the file structure at dstPath
+func Untgz(dstPath string, r io.Reader) error {
+
+	gzr, err := gzip.NewReader(r)
+	if err != nil {
+		return fmt.Errorf("error reading file: %w", err)
+	}
+	defer gzr.Close()
+
+	tr := tar.NewReader(gzr)
+
+	for {
+		header, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("error while iterating on tar reader: %w", err)
+		}
+		if header == nil {
+			continue
+		}
+
+		target := filepath.Join(dstPath, header.Name)
+
+		// Sanity check to protect against zip-slip
+		if !strings.HasPrefix(target, filepath.Clean(dstPath)+string(os.PathSeparator)) {
+			return fmt.Errorf("illegal filepath in archive: %s", target)
```

Comment:
Can we avoid returning the full path? I think the first part of the path should be considered secret.

It's possible that an error message sanitization interceptor would solve that problem generally.